src/message.rs

```diff
@@ -0,0 +1,337 @@
+use pyo3::{
+    exceptions::{PyNotImplementedError, PyValueError},
+    pyclass,
+    pyclass_init::PyObjectInit,
+    pymethods,
+    types::{PyBytes, PyString, PyType},
+    PyClassInitializer, PyObject, PyRef, PyResult, PyTypeInfo, Python,
+};
+
+use crate::{
+    api_crypto::{PrivateKey, PublicKey, SecretKey, SigningKey, VerifyKey},
+    ids::{DeviceID, EntryID},
+    manifest::EntryName,
+    time::DateTime,
+};
+
+#[pyclass(subclass)]
+#[derive(Clone)]
+pub(crate) struct MessageContent(pub libparsec::types::MessageContent);
+
+crate::binding_utils::gen_proto!(MessageContent, __repr__);
+crate::binding_utils::gen_proto!(MessageContent, __richcmp__, eq);
+
+#[pymethods]
+impl MessageContent {
+    #[getter]
+    fn author(&self) -> PyResult<DeviceID> {
+        Ok(match &self.0 {
+            libparsec::types::MessageContent::SharingGranted { author, .. } => {
+                DeviceID(author.clone())
+            }
+            libparsec::types::MessageContent::SharingReencrypted { author, .. } => {
+                DeviceID(author.clone())
+            }
+            libparsec::types::MessageContent::SharingRevoked { author, .. } => {
+                DeviceID(author.clone())
+            }
+            libparsec::types::MessageContent::Ping { author, .. } => DeviceID(author.clone()),
+        })
+    }
+
+    #[getter]
+    fn timestamp(&self) -> PyResult<DateTime> {
+        Ok(match self.0 {
+            libparsec::types::MessageContent::SharingGranted { timestamp, .. } => {
+                DateTime(timestamp)
+            }
+            libparsec::types::MessageContent::SharingReencrypted { timestamp, .. } => {
+                DateTime(timestamp)
+            }
+            libparsec::types::MessageContent::SharingRevoked { timestamp, .. } => {
+                DateTime(timestamp)
+            }
+            libparsec::types::MessageContent::Ping { timestamp, .. } => DateTime(timestamp),
+        })
+    }
+
+    #[classmethod]
+    pub fn decrypt_verify_and_load_for(
+        _cls: &PyType,
+        ciphered: &[u8],
+        recipient_privkey: &PrivateKey,
+        author_verify_key: &VerifyKey,
+        expected_author: &DeviceID,
+        expected_timestamp: DateTime,
+        py: Python,
+    ) -> PyResult<PyObject> {
+        let msg = libparsec::types::MessageContent::decrypt_verify_and_load_for(
+            ciphered,
+            &recipient_privkey.0,
+            &author_verify_key.0,
+            &expected_author.0,
+            expected_timestamp.0,
+        );
+
+        Ok(unsafe {
```

Comment:
No way to remove the use of `unsafe` ?

Reply:
Nope if we work with subclass

Reply:
Add a comment to explain why this block is safe, I plan to enable [`clippy::undocumented_unsafe_blocks`](https://rust-lang.github.io/rust-clippy/master/#undocumented_unsafe_blocks).
Can you also limit the size of the unsafe block to code that required ?

i.e.: only put `into_new_object` & `from_owned_ptr` into the `unsafe` block and keep everything else outside

Reply:
:+1: